core/src/main/java/com/sequenceiq/cloudbreak/rotation/context/provider/UserKeyPairRotationContextProvider.java

```diff
@@ -0,0 +1,208 @@
+import com.sequenceiq.cloudbreak.rotation.secret.RotationContext;
+import com.sequenceiq.cloudbreak.rotation.secret.RotationContextProvider;
+import com.sequenceiq.cloudbreak.rotation.secret.SecretType;
+import com.sequenceiq.cloudbreak.rotation.secret.step.CommonSecretRotationStep;
+import com.sequenceiq.cloudbreak.rotation.secret.step.SecretRotationStep;
+import com.sequenceiq.cloudbreak.service.GatewayConfigService;
+import com.sequenceiq.cloudbreak.service.environment.EnvironmentClientService;
+import com.sequenceiq.cloudbreak.service.environment.credential.CredentialConverter;
+import com.sequenceiq.cloudbreak.service.stack.StackDtoService;
+import com.sequenceiq.cloudbreak.view.InstanceMetadataView;
+import com.sequenceiq.environment.api.v1.credential.model.response.CredentialResponse;
+import com.sequenceiq.environment.api.v1.environment.model.response.DetailedEnvironmentResponse;
+import com.sequenceiq.environment.api.v1.environment.model.response.EnvironmentAuthenticationResponse;
+
+@Component
+public class UserKeyPairRotationContextProvider implements RotationContextProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserKeyPairRotationContextProvider.class);
+
+    private static final int MAX_RETRY = 10;
+
+    private static final String REPLACE_SSH_PUBLICKEY_STATE = "ssh.replace_ssh_publickey";
+
+    @Inject
+    private StackDtoService stackService;
+
+    @Inject
+    private EnvironmentClientService environmentClientService;
+
+    @Inject
+    private CloudPlatformConnectors cloudPlatformConnectors;
+
+    @Inject
+    private CredentialToCloudCredentialConverter cloudCredentialConverter;
+
+    @Inject
+    private CredentialConverter credentialConverter;
+
+    @Inject
+    private RegionAwareInternalCrnGeneratorFactory regionAwareInternalCrnGeneratorFactory;
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private StackAuthenticationRepository stackAuthenticationRepository;
+
+    @Override
+    public Map<SecretRotationStep, RotationContext> getContexts(String resourceId) {
+        Map<SecretRotationStep, RotationContext> result = Maps.newHashMap();
+        StackDto stack = stackService.getByCrn(resourceId);
+        DetailedEnvironmentResponse environment = ThreadBasedUserCrnProvider.doAsInternalActor(
+                regionAwareInternalCrnGeneratorFactory.iam().getInternalCrnForServiceAsString(),
+                () -> environmentClientService.getByCrn(stack.getEnvironmentCrn()));
+
+        boolean changed = customerChangedTheKeyPair(stack.getStackAuthentication(), environment.getAuthentication());
+
+        result.put(CloudbreakSecretRotationStep.SALT_STATE_RUN, getSaltRunOrchestratorStateRotationContext(changed, stack.getResourceCrn(), stack, environment));
+        result.put(CommonSecretRotationStep.CUSTOM_JOB, getCustomJobRotationContext(changed, stack.getResourceCrn(), stack, environment));
```

Comment:
maybe worth to extract these two methods into separate classes or at least one, the class looks big